feat: add NetworkPolicy template

GabboPenna opened the pull request:
### Proposed changes

This PR adds optional NetworkPolicy support to the nginx-ingress Helm chart.

It introduces:

- A new `controller.networkPolicy` section in `values.yaml` with:
  - An `enabled` flag.
  - Configurable `ingress` and `egress` rules.
- A NetworkPolicy template that always derives its `podSelector` from `nginx-ingress.selectorLabels` so it targets the correct controller pods.
- Automatic `policyTypes` generation based on whether ingress and/or egress rules are configured.
- JSON schema validation and Helm snapshot coverage.

With this change, chart consumers can declare fine-grained network controls for their Ingress Controller pods without maintaining external manifests.

Fixes: #8060

### Checklist

- [x] I have read the CONTRIBUTING doc
- [x] I have added tests that prove my feature works
- [x] I have checked that all Helm unit tests and `helm lint` pass
- [x] I have updated the necessary values documentation and schema
- [x] I have refreshed the branch on `main`
- [x] I have hereby read the F5 CLA and agree to its terms

GabboPenna committed:
feat: add NetworkPolicy template

AlexFenlon commented:
Hi @GabboPenna,

Thanks for this PR, 

Can you please create an issue for this so we can track it appropriately.

GabboPenna commented:
> Hi @GabboPenna,
> 
> Thanks for this PR,
> 
> Can you please create an issue for this so we can track it appropriately.

Hi @AlexFenlon, i open a new issue: [#8060](https://github.com/nginx/kubernetes-ingress/issues/8060)

vepatel committed:
Merge branch 'main' into add-networkpolicy-support

vepatel commented:
@hi @GabboPenna, can you please also update helm schema files for NetworkPolicy: https://github.com/GabboPenna/kubernetes-ingress/blob/add-networkpolicy-support/charts/nginx-ingress/values.schema.json

vepatel committed:
Merge branch 'main' into add-networkpolicy-support

GabboPenna committed:
Update netpool template and values

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GabboPenna committed:
testdata for netpool and new snap

GabboPenna committed:
feat: add configurable controller NetworkPolicy

GabboPenna committed:
Merge previous NetworkPolicy work into refreshed branch

GabboPenna commented:
Hi @haywoodsh, apologies for the long delay. I agree with your suggestions and have updated the PR accordingly.

The `podSelector` is now derived from `nginx-ingress.selectorLabels`, while `policyTypes` is automatically generated from the configured ingress and egress rules. I also refreshed the implementation on the current `main`, updated the schema, and added Helm test coverage and snapshots.

I verified the changes locally with Helm v3.21.2 using the full chart test suite and `helm lint`.

Thanks for reopening the PR and for your patience. Please let me know if anything else needs adjusting.

AlexFenlon commented:
Hi @GabboPenna, thanks for coming back to this PR! great to see you again. 

we really appreciate that you addressed the feedback, we will look into this as soon as we can!

GabboPenna committed:
docs(helm): clarify empty NetworkPolicy rules

GabboPenna commented on charts/nginx-ingress/values.yaml:
```diff
@@ -714,6 +714,15 @@ controller:
+  networkPolicy:
+    ## Create a NetworkPolicy that targets the controller pods.
+    enabled: false
+    ## Configure ingress rules. When non-empty, the Ingress policy type is added automatically.
+    ingress: []
+    ## Configure egress rules. When non-empty, the Egress policy type is added automatically.
+    egress: []
```
Good point. I added an explicit warning in values.yaml that enabling the NetworkPolicy with both rule lists empty denies all ingress by default.

pdabelf5 commented:
/approve-pipeline-run

vepatel commented:
/approve-pipeline-run
